service_level_objectives.go

```diff
@@ -0,0 +1,360 @@
+		if i != 0 && name != "" {
+			result[name] = match[i]
+		}
+	}
+	if len(result) != 2 {
+		return 0, fmt.Errorf("invalid timeframe specified: '%s'", timeframe)
+	}
+	qty, err := json.Number(result["quantity"]).Int64()
+	if err != nil {
+		return 0, fmt.Errorf("invalid timeframe specified, could not convert quantity to number")
+	}
+	if qty <= 0 {
+		return 0, fmt.Errorf("invalid timeframe specified, quantity must be a positive number")
+	}
+
+	switch result["unit"] {
+	// FUTURE: will support more time frames, hence the switch here.
+	default:
+		// only matches on `d` currently, so this is simple
+		return time.Hour * 24 * time.Duration(qty), nil
+	}
+}
+
+// CreateServiceLevelObjective adds a new service level objective to the system. This returns a pointer
+// to the service level objective so you can pass that to UpdateServiceLevelObjective or DeleteServiceLevelObjective
+// later if needed.
+func (client *Client) CreateServiceLevelObjective(slo *ServiceLevelObjective) (*ServiceLevelObjective, error) {
+	var out reqServiceLevelObjectives
+
+	if slo == nil {
+		return nil, fmt.Errorf("no SLO specified")
+	}
+
+	if err := client.doJsonRequest("POST", "/v1/slo", slo, &out); err != nil {
+		return nil, err
+	}
+	if out.Error != "" {
+		return nil, fmt.Errorf(out.Error)
+	}
+	if len(out.Data) != 1 {
+		return nil, fmt.Errorf("not found")
+	}
+
+	return out.Data[0], nil
+}
+
+// UpdateServiceLevelObjective takes a service level objective that was previously retrieved through some method
+// and sends it back to the server.
+func (client *Client) UpdateServiceLevelObjective(slo *ServiceLevelObjective) error {
+	var out reqServiceLevelObjectives
+
+	if slo == nil {
+		return fmt.Errorf("no SLO specified")
+	}
+
+	if err := client.doJsonRequest("PUT", fmt.Sprintf("/v1/slo/%s", slo.GetID()), slo, &out); err != nil {
+		return err
+	}
+
+	if out.Error != "" {
+		return fmt.Errorf(out.Error)
+	}
+
+	if len(out.Data) != 1 {
+		return fmt.Errorf("not found")
+	}
+
+	*slo = *out.Data[0]
+
+	return nil
+
+}
+
+type reqServiceLevelObjectives struct {
+	Data  []*ServiceLevelObjective `json:"data"`
+	Error string                   `json:"error"`
+}
+
+// SearchServiceLevelObjectives searches for service level objectives by search criteria.
+// limit will limit the amount of SLO's returned, the API will enforce a maximum and default to a minimum if not specified
+func (client *Client) SearchServiceLevelObjectives(limit int, offset int, query string, ids []string) ([]*ServiceLevelObjective, error) {
+	var out reqServiceLevelObjectives
+	uriValues := make(url.Values, 0)
+	if limit > 0 {
+		uriValues.Set("limit", fmt.Sprintf("%d", limit))
+	}
+	if offset >= 0 {
+		uriValues.Set("offset", fmt.Sprintf("%d", offset))
+	}
+	// specify by query
+	if query != "" && len(ids) == 0 {
+		uriValues.Set("query", query)
+	}
+	// specify by `ids`
+	if query == "" && len(ids) > 0 {
+		uriValues.Set("ids", strings.Join(ids, ","))
+	}
```

Comment:
So what should happen if `query != "" && len(ids) >0`? (I'm assuming that if `query == "" && len(ids) == 0`, then nothing needs to be done)?

Reply:
Correct, both are optional

Reply:
fixing this inline docs and better organization